### PR TITLE
Feature/migerate to terraform

### PR DESCRIPTION
**Pull request by mrmasterplan**
# Ready to merge. Please review.
---

## What type of PR is this?

- Refactor
- Bug Fix
- Optimization
- Deployment (Azure, CI/CD, PowerShell, ...)

## Description

- Introduce a test deployment with full Unity Catalog enabled
- Test still create Databases and tables in HIVE, but UC is fully available. Future PRs should test it more.
- Change the spetlr version to be the same as the Databricks Runtime Version used in testing - 14.3.* for now
- Test only on 1 DBR version
- Make Speltr UC compatible (see below)
- Change the test deployment to be completely in terraform
- SQL server test were not ported to terraform yet, we invite future PRs to re-introduce this part of the tests.
- 

### Functional changes to Speltr

- The DeltaHandle by default no longer reads or writes by path. All read and writes are done using the name. 
         Unity Catalog tables are either managed, residing in UC managed storage or external (i.e. residing in storage that is known to UC, but not exposed to the user.) In both cases UC wants the user to go trough UC when accessing the table so that permissions can be applied. Therefore all direct access to the underlying data (regardless if that may save a few milliseconds) is not allowed.
- SQL server classes and some other code areas have been marked as deprecated for the simple reason that they are untested. The code has not been changed from previous versions and can still be used.
- Azure tag access is just a convenience method to access azure workspace resource tags in code.
- The Version has been increased to 14.3.0 to align with the DBR version that we test with.

### Deployment

Deployment is split into several stages:
- Permanent resources (metastore_and_pemanent) deploys the metastore and the permanent resources (Cosmos account)
- "integration_resources" deploys all azure resources, the captain service principal and the workspace.
- "integration_databricks" then uses that workspace API and deploys the catalog, workspace association, external locations and permissions so that the SPN can deploy the test job
- Test job submission is a simple step. The spetlr-tools repos could in fact use the pipeline to here and deploy its own set of commands at this point.

## Next steps after this

- Sql Server tests were taken out to descope the transformation. The powershell code has been left in the repo to serve as a reference for when these test resources are ported into terraform.
- Create a cleanup pipeline that deletes the captains and catalogs from databricks. Currently the tests leak 1 catalog and 1 databricks service principal on each run. They can be deleted by administrators. A cleanup script is needed but not critical.

### CODE REVIEW

**Comment by mrmasterplan** [comment]
Hi, @FarBo, I closed your PR and opened a new one so that you are able to add an approval to my work.

**Comment by FarBo** [comment]
I will take a look today. Thanks for the work

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ jobs:
         with:
           fetch-depth: 0
       - name: Set up Python
-        uses: actions/setup-python@v2
+        uses: actions/setup-python@v5
         with:
           python-version: "3.8"
```
Comment (FarBo) [inline .github/workflows/release.yml:18]: increase to 3.10 here as well

File: .github/terraform/integration_databricks/data_and_varaibles.tf [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+data "azurerm_key_vault_secret" "captain_spn_tenant" {
+  name         = module.config.integration.captain.kv_secret_tenant
+  key_vault_id = data.azurerm_key_vault.key_vault.id
+}
+
+data "databricks_metastore" "db_metastore" {
+  provider = databricks.account
+
```
Comment (FarBo) [inline .github/terraform/integration_databricks/data_and_varaibles.tf:36]: remove space

File: .github/deploy/Utilities/Graph/Graph-CreateRole.ps1
Comment (FarBo) [inline .github/deploy/Utilities/Graph/Graph-CreateRole.ps1:1]: Some of these graph modules are used in creating cicd spn script: setup_deploying_SPN.ps1

File: .github/terraform/integration_databricks/data_and_varaibles.tf (hunk 2 above)
Comment (mrmasterplan) [reply .github/terraform/integration_databricks/data_and_varaibles.tf:36]: Not sure which space you mean. I use `terraform fmt` so I will not change spacing inside lines. I have consistently separated the provider with a newline everywhere, so I also want to keep that.

File: .github/deploy/Utilities/Graph/Graph-CreateRole.ps1
Comment (FarBo) [reply .github/deploy/Utilities/Graph/Graph-CreateRole.ps1:1]: Let's create and issue for fixing the cicd script after these deletions

File: .github/submit/main.py
Comment (FarBo) [inline .github/submit/main.py:1]: This can be removed as discussed

**Comment by JeppeBlixen** [comment]
A couple of points
- I think this PR is too lange. I have no change reviewing 100 commits on 150 changed files. Reading the description it also feels like different features, that could be in separate PR's.
- Just deprecating the sql server class does not sound good. I know it was used on a customer, where we stopped halv a year ago.
- I don't like the braking change of the DeltaHandle. Personally i agree that the read should be on table name by default, but there are situations where the location is used. I think we should remove the table name as required patrameter, and use location if table name is not supplied.

**Comment by FarBo** [comment]
Let's check if schema and table creations end up in the catalog as we expect or hive_metastore.

**Review by FarBo — APPROVED** [review]
I think this can be merged in the current state. If the test catalogs and service principals can be deleted, it would be great to add it here, otherwise it can also be added later.

Since, the test data are not assigned in the created catalog in unity, and still point to the legacy hive_metastore, I would consider changing the description of the PR to reflect that.

**Comment by mrmasterplan** [comment]
> A couple of points
> 
> * I think this PR is too lange. I have no chance reviewing 100 commits on 150 changed files. Reading the description it also feels like different features, that could be in separate PR's.
> * Just deprecating the sql server class does not sound good. I know it was used on a customer, where we stopped halv a year ago.
> * I don't like the braking change of the DeltaHandle. Personally i agree that the read should be on table name by default, but there are situations where the location is used. I think we should remove the table name as required patrameter, and use location if table name is not supplied.

Good points! Some replies:
- I did not change the SqlServer class. I simply removed the tests. We will reintroduce the tests in future PRs precisely because this PR is already too large.
- The size of this PR is because it is a fundamental change. If we had multiple release branches, we could run this new form in parallel. OR we could fork the repo, but we prefer to keep the main branch at the latest stage. All users are welcome to wait with updating to this new version until they are happy with the state.
- Regarding the breaking change in delta handle, I will think about it more and get back to you.

**Comment by mrmasterplan** [comment]
@JeppeBlixen I have added code to restore the delta-handle's ability to use only the path with no name

**Pull request merged**